lib/backends/webgl/glsl-coordinate-lib.ts

```diff
@@ -54,6 +71,1072 @@ export class CoordsGlslLib extends GlslLib {
+            })
+            .join('');
+
+    source = `
+      ivec5 getOutputCoords() {
+          ivec2 resTexRC = ivec2(TexCoords.yx *
+                                vec2(${texShape[0]}, ${texShape[1]}));
+          int index = resTexRC.y * ${texShape[1]} + resTexRC.x;
+          ${coordsFromIndexSnippet}
+          return ivec5(r, c, d, d2, d3);
+        }
+      `;
+    return new GlslLibRoutine(source);
+  }
+
+  /**
+   * ND output coordinates.
+   */
+  protected getOutputPackedNDCoords(shape: ReadonlyArray<number>, texShape: [number, number]): GlslLibRoutine {
+    const packedTexShape = [Math.ceil(texShape[0] / 2), Math.ceil(texShape[1] / 2)];
+
+    const texelsInLogicalRow = Math.ceil(shape[shape.length - 1] / 2);
+    const texelsInBatch = texelsInLogicalRow * Math.ceil(shape[shape.length - 2] / 2);
+    let texelsInBatchN = texelsInBatch;
+    let batches = ``;
+    let coords = 'b, r, c';
+
+    for (let b = 2; b < shape.length - 1; b++) {
+      texelsInBatchN *= shape[shape.length - b - 1];
+      batches = `
+      int b${b} = index / ${texelsInBatchN};
+      index -= b${b} * ${texelsInBatchN};
+    ` + batches;
+      coords = `b${b}, ` + coords;
+    }
+    const source = `
+      ivec${shape.length} getOutputCoords() {
+        ivec2 resTexRC = ivec2(TexCoords.xy *
+                              vec2(${packedTexShape[0]}, ${packedTexShape[1]}));
+        int index = resTexRC.y * ${packedTexShape[1]} + resTexRC.x;
+
+        ${batches}
+
+        int b = index / ${texelsInBatch};
+        index -= b * ${texelsInBatch};
+
+        // reverse r and c order for packed texture
+        int r = imod(index, ${texelsInLogicalRow}) * 2;
+        int c = 2 * (index / ${texelsInLogicalRow});
+
+        return ivec${shape.length}(${coords});
+      }
+    `;
+    return new GlslLibRoutine(source);
+  }
+
+  /**
+   * Generates code for common UV coords computation utility functions.
+   */
+  protected GetCommonUtilFuncs(): {[name: string]: GlslLibRoutine;} {
+    const result: {[name: string]: GlslLibRoutine} = {};
+    let funcName = `uvFromFlat`;
+    result[funcName] = new GlslLibRoutine(`
+    vec2 uvFromFlat(int texNumR, int texNumC, int index) {
+      int texC = index / texNumR;
+      int texR = index - texC * texNumR;
+      // TODO: swap texR, texC order in following function so row is corresponding to u and column is corresponding to v.
+      return (vec2(texR, texC) + halfCR) / vec2(texNumR, texNumC);
+    }
+    `);
+    funcName = `packedUVfrom1D`;
+    result[funcName] = new GlslLibRoutine(`
+      vec2 packedUVfrom1D(int texNumR, int texNumC, int index) {
+        int texelIndex = index / 2;
+        int texR = texelIndex / texNumC;
+        int texC = texelIndex - texR * texNumC;
+        return (vec2(texC, texR) + halfCR) / vec2(texNumC, texNumR);
+      }
+      `);
+    funcName = `packedUVfrom2D`;
+    result[funcName] = new GlslLibRoutine(`
+      vec2 packedUVfrom2D(int texelsInLogicalRow, int texNumR,
```

Comment:
align argument list with other functions, so it would be
vec2 packedUVfrom2D(int texNumR, int texNumC, int texelsInLogicalRow, int row, int col) ?